apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/supervisor/client/impl/MessageDispatcherImpl.java

```diff
@@ -0,0 +1,270 @@
+    ExecutorService executor = Executors.newCachedThreadPool();
+    private final Map<Long, ResponseLatch> waiting = new ConcurrentHashMap<>();
+
+    public MessageDispatcherImpl() {
+        pathMatcher = new PathParamProcessor();
+        connections = new SvConnections(this);
+    }
+
+    public Map<URI, SvBusClient> getConnections() {
+        return connections.getAll();
+    }
+
+    /**
+     *
+     * @return request handler that matches path specification
+     */
+    private HandlerRecord findRqByPath(String path) {
+        HandlerRecord handlerRec = pathMatcher.findAndParse(path);
+        return handlerRec;
+    }
+
+    @Override
+    public void registerRqHandler(String pathSpec, Class<? extends SvBusMessage> rqMapping, Class<? extends SvBusMessage> respMapping, SvRequestHandler handler) {
+        pathMatcher.registerRqHandler(pathSpec, rqMapping, respMapping, handler);
+    }
+
+    @Override
+    public void unregisterRqHandler(String pathSpec) {
+        pathMatcher.remove(pathSpec);
+    }
+
+    long nextMessageId() {
+        return Math.round(Math.random() * (Long.MAX_VALUE - 1));
+    }
+
+    boolean isMyAddress(String to) {
+        return myAddress.equals(to);
+    }
+
+    public SvBusResponse errornousRequestsHandler(JsonNode rqBody, SvChannelHeader rqHeader, int code, String errorInfo) {
+        SvBusResponse resp = new SvBusResponse();
+        resp.error = new SvBusError();
+        resp.error.error_code = code;
+        resp.error.descritption = errorInfo + " Request path: " + rqHeader.path;
+        return resp;
+    }
+
+    void addConnection(URI uri, boolean isDefault) {
+        SvBusClient client = new SvBusClient(uri, this);
+        connections.put(uri, client, isDefault);
+        client.connect();
+    }
+
+    private SvBusResponse getResponse(Long rqId) throws SocketTimeoutException {
+        SvBusResponse res = null;
+        ResponseLatch l = waiting.get(rqId);
+        if (l != null) {
+            res = l.get(RESPONSE_WAIT_TIMEOUT_MS);
+        }
+        return res;
+    }
+
+    public SvBusResponse sendSync(SvBusMessage rq, String path, URI addr) {
+        SvBusResponse res = null;
+        SvBusClient client = connections.get(addr);
+        if (client == null) {
+            client = connections.getDefault().getValue();
+        }
+        SvChannelHeader header = new SvChannelHeader();
+        header.from = myAddress.toString();
+        header.to = addr.toString();
+        header.path = path;
+        header.messageId = nextMessageId();
+        SvChannelMessage env = new SvChannelMessage();
+        env.header = header;
+        env.body = rq;
+        try {
+            if (!client.sendMessage(env)) {
+                outgoingQueue.addLast(env);
+            }
+            waiting.put(header.messageId, new ResponseLatch());
+            res = getResponse(header.messageId);
+
+        } catch (SocketTimeoutException ex) {
+            log.warn("Responce wait timeout", ex);
```

Comment:
in this case, it's worth to return a response with a suitable error code and description instead of null-response;

Reply:
I think that event better to propagate error to the caller code